Add a unique ID to Interval properties, allow multiple intervals with the same (start, end)

pleath opened the pull request:
Add methods to get and remove an interval from a collection by ID. Use ID as a third key to determine order in cases of identical (start, end). "delete(start, end)" will remove all (start, end) intervals from the collection. Each such delete will be serialized as an independent operation.

pleath committed:
Add methods to get and remove an interval from a collection by ID

Add and delete ops will be supplied with ID's. delete(start, end) will be deprecated; for now it will delete the interval with a legacy ID based on start/end.

pleath commented:
Replaces https://github.com/microsoft/FluidFramework/pull/6279.

pleath commented:
Addresses #6108, #6109.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -677,52 +735,52 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
-            sequenceNumber = this.client.getCurrentSeq();
+        end: number,
+        local: boolean = true,
+        op?: ISequencedDocumentMessage) {
```
this doesn't look right. we shouldn't be exposing op or local on public methods

anthony-murphy replied:
this goes for a number of methods. the public api surface should be abstracted from the internals of dealing with remote ops.

pleath replied:
I see. I'll refactor the delete path so this doesn't happen. You're not talking about, for instance, 


<!--StartFragment-->public addInternal(        serializedInterval: ISerializedInterval, local: boolean, op: ISequencedDocumentMessage)<!--EndFragment-->

, right? For LocalCollection.addInterval, I passed in "local" to control whether we create a new ID. If we got an interval over the wire without an ID, I thought it would be best to leave it in that state, rather than have each client create its own ID. But maybe that's not a case we need to worry about, so maybe LocalCollection.addInterval can just create an ID if it doesn't see one in the properties.

anthony-murphy commented:
>         it("propagates", async () => {

can you add some multi client tests

---
Refers to: packages/test/test-end-to-end-tests/src/test/sharedInterval.spec.ts:337 in aa65640. [](commit_id = aa656409b2466d834de50a8fbad920c314eec614, deletion_comment = False)

anthony-murphy commented:
>     a.addPropertySet(b.properties);

how does this, defaultIntervalConflictResolver behave with local vs remote conflict? do both clients end up with the same id, or will different id's never conflict

---
Refers to: packages/dds/sequence/src/intervalCollection.ts:288 in aa65640. [](commit_id = aa656409b2466d834de50a8fbad920c314eec614, deletion_comment = False)

anthony-murphy commented:
>         it("propagates", async () => {

it would be good to add some scenario focus tests. currently you really only test the local client, as there is only one container, and you never load a new container to test rehydration

---
In reply to: [861874250](https://github.com/microsoft/FluidFramework/pull/6407#issuecomment-861874250) [](http://example.com/codeflow?ancestors=861874250)

---
Refers to: packages/test/test-end-to-end-tests/src/test/sharedInterval.spec.ts:337 in aa65640. [](commit_id = aa656409b2466d834de50a8fbad920c314eec614, deletion_comment = False)

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -149,7 +156,21 @@ export class SequenceInterval implements ISerializableInterval {
+                    const bId = b.getIntervalId();
+                    if (bId) {
+                        return thisId > bId ? 1 : thisId < bId ? -1 : 0;
+                    }
```
do we explcitly support adding a interval with the same id as and existing interval for update?

pleath replied:
Well, we don't have an API that lets you specify the ID when you add an interval (though you could mess with an interval's ID if you wanted to once it's been created). The local add(start, end) will create a new ID, and the non-local add(serialized interval) will propagate the ID from the serialized interval. (I changed the flow so the local add(start, end) no longer gets routed through add(serialized interval).) For the case where you want to update an existing interval, I'm picturing a new API like change(ID, start, end). For now, I'm having it do a remove-update-add under the hood, but I know we want to send a single op over the wire.

pleath commented:
https://github.com/microsoft/FluidFramework/pull/6407#issuecomment-861874250

I do see a client receiving the relevant ops in the test I've been adding to. At least the non-local add(serialized) and delete(serialized) getting called to receive the ops. Can you point me to an example of code that loads a new container to test rehydration?

pleath commented:
https://github.com/microsoft/FluidFramework/pull/6407#issuecomment-861876536

Yeah, different ID's won't conflict. The compare function will compare ID's if (start, end) are the same.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -722,7 +766,8 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
-                this.emitter.emit("add", undefined, serializedInterval);
+                // Review: Is this case possible?
+                this.emitter.emit("delete", undefined, serializedInterval);
             } else {
```
i don't understand this case? why send a delete op?

pleath replied:
Oh, that just looks like a bad merge. I'll fix it.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -406,6 +435,12 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
         }
     }
 
+    public findInterval(start: number, end: number) {
```
does this return an array, or a single interval?

anthony-murphy replied:
this method is odd. which interval does it return? with or without id? or either?

pleath replied:
It returns a single interval, basically, the first one it finds that matches (start, end). Yeah, maybe that's odd, and it would be nicer to return an array. Then, for the delete(start, end) case, the caller could walk the array doing the individual deletes.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -441,19 +481,33 @@ export class LocalIntervalCollection<TInterval extends ISerializableInterval> {
         intervalType: MergeTree.IntervalType,
-        props?: MergeTree.PropertySet) {
-        const interval = this.createInterval(start, end, intervalType);
+        props: MergeTree.PropertySet,
```
>props [](http://example.com/codeflow?start=8&length=5)

why make this not optional?

anthony-murphy replied:
ahh this is the local collection, so it's internal only, hard to keep all the classes straight in this file

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -677,43 +735,29 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
+                end,
+                intervalType,
+                properties: interval.properties,
+                sequenceNumber: this.client?.getCurrentSeq() ?? 0,
```
>sequenceNumber [](http://example.com/codeflow?start=16&length=14)

we would usually call this referenceSequenceNumber, as sequenceNumbers are provided by the server

anthony-murphy replied:
looks to be pre-existing

pleath replied:
Do you think the member name in ISerializedInterval should be changed?

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -677,43 +735,29 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
     // TODO: error cases
-    public addInternal(serializedInterval: ISerializedInterval, local: boolean, op: ISequencedDocumentMessage) {
-        const interval = this.localCollection.addInterval(
+    public addInternal(
```
>addInternal [](http://example.com/codeflow?start=11&length=11)

should the above add method call this

anthony-murphy replied:
looks like you moved some methods which make the merge gross, but i'm not sure why anything in add needed to change

anthony-murphy replied:
add used to call addInternal, which would hit the case below

pleath replied:
I changed it to make the local and non-local cases more distinct. I had to do that for delete. It does make debugging easier if addInternal is only called in the case where an op comes over the wire. Sorry about the merge, though.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -735,23 +780,62 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
 
-    public deleteInterval(serializedInterval: ISerializedInterval, local: boolean, op: ISequencedDocumentMessage) {
-        const interval = this.localCollection.removeInterval(serializedInterval.start, serializedInterval.end);
+    public delete(start: number, end: number) {
```
>delete [](http://example.com/codeflow?start=11&length=6)

i don't think this is the right behavior. i don't think we should ever loop in delete. i see two cases, delete the interval at the position without id, or with id. i think both these cases should be explicit

pleath replied:
The behavior we decided on was that delete(start, end) would delete all the intervals that match (start, end), and one delete op would go over the wire (with ID) for each interval. So if client A does delete(start, end), it looks to client B as though client A did delete(id1), delete(id2), ... . That's what this is doing, but I do think it would be nicer to have findInterval return an array and then have delete(start, end) loop over the array.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -735,23 +780,62 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
+        const id = serializedInterval.properties?.[reservedIntervalIdKey];
+        if (id === undefined) {
+            // Delete all (start, end).
+            const start = serializedInterval.start;
```
given above, which will send an delete op per interval, this seems wrong as well

anthony-murphy replied:
it looks like we will send and op per interval at a range, and then delete all intervals for each op. as i mention above, i dislike this type of behavior, as it is hard to get eventually consistent. i would make operations granular, specifcally 1 op per interval

pleath replied:
Yeah, the ops over the wire are granular. In the delete path, only deleteExistingInterval sends out a serialized op.

anthony-murphy commented on packages/dds/sequence/src/intervalCollection.ts:
```diff
@@ -735,23 +790,52 @@ export class IntervalCollectionView<TInterval extends ISerializableInterval> ext
+     * @deprecated - delete(start,end) will be removed in future.  Please use deleteIntervalById instead.
+     */
+    public delete(start: number, end: number) {
+        return this.deleteIntervalById(this.localCollection.createLegacyId(start, end));
```
it doesn't actually work this way. you need to actually set the reserved id field in props on load from the snapshot, as in the case of shared inverval on a shared string the endpoints can change over time as modifications happen. 

overall, i think we might be tackling too many things is this pr, which is making it hard to get any of them in.

pleath replied:
Updating to handle the load case, which is the one I wasn't handling.

pleath committed:
Add dehydrate+rehydrate test

pleath committed:
Allow multiple intervals with the same (start, end) in a collection

Use the interval ID as a third key for sorting. Fix an issue in serialization of add ops that was causing the original ID not to be sent over the wire.

pleath committed:
Fix behavior of delete(start, end) and a cross-client issue with transmission of ID's

anthony-murphy reviewed (approved):
:shipit:

pleath committed:
Restore global uniqueness of legacy ID's
